Zypper conveyorPacker

ikaneshiro opened the pull request:
Conveyor Packer that allows for bootstrapping with zypper.

fixes #1944

ikaneshiro committed:
zypper conveyorPacker

ikaneshiro committed:
make  escalate warning to error and clean up redundant sylog.error()

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE file distributed with the sources of this project regarding your
```
.md

ikaneshiro committed:
LICENSE -> LICENSE.md

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+		return fmt.Errorf("zypper is not in PATH: %v", err)
+	}
+
+	// //check for rpm on system
```
double //

ArangoGutierrez replied:
and `//` + space

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+
+	include, _ := cp.b.Recipe.Header["include"]
+
+	//check for include environment variable and add it to requires string
```
`//` + space

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+	//check for include environment variable and add it to requires string
+	include += ` ` + os.Getenv("INCLUDE")
+
+	//trim leading and trailing whitespace
```
`//` + space

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+	//trim leading and trailing whitespace
+	include = strings.TrimSpace(include)
+
+	//add aaa_base to start of include list by default
```
`//` + space

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+	//add aaa_base to start of include list by default
+	include = `aaa_base ` + include
+
+	//Create the main portion of zypper config
```
`//` + space

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+		return fmt.Errorf("While copying pseudo devices: %v", err)
+	}
+
+	//Add mirrorURL as repo
```
`//` + space

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+		return fmt.Errorf("While adding zypper mirror: %v", err)
+	}
+
+	//Refreshing gpg keys
```
`//` + space

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+	args := []string{`--non-interactive`, `-c`, filepath.Join(cp.b.Rootfs(), zypperConf), `--root`, cp.b.Rootfs(), `--releasever=` + osversion, `-n`, `install`, `--auto-agree-with-licenses`}
+	args = append(args, strings.Fields(include)...)
+
+	//Zypper install command
```
`//` + space

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+
+	sylog.Debugf("\n\tZypper Path: %s\n\tDetected Arch: %s\n\tOSVersion: %s\n\tMirrorURL: %s\n\tIncludes: %s\n", zypperPath, runtime.GOARCH, osversion, mirrorurl, include)
+
+	//run zypper
```
`//` + space

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+
+// Pack puts relevant objects in a Bundle!
+func (cp *ZypperConveyorPacker) Pack() (b *types.Bundle, err error) {
+
```
white line

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+}
+
+func (cp *ZypperConveyorPacker) genZypperConfig() (err error) {
+
```
white line

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+		return fmt.Errorf("Invalid zypper header, no MirrorURL specified")
+	}
+
+	//look for an OS version if the mirror specifies it
```
`//` + space

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,252 @@
+		return
+	}
+
+	//get mirrorURL, OSVerison, and Includes components to definition
```
`//` + space

ikaneshiro committed:
code style changes

ikaneshiro committed:
removing unneeded concurrency

ikaneshiro committed:
Merge branch 'master' of github.com:sylabs/singularity into zypper_dev_manual

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper_test.go:
```diff
@@ -0,0 +1,96 @@
+const zyppDef = "../../../../examples/opensuse/Singularity"
+
+func TestZypperConveyor(t *testing.T) {
+
```
white line

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper_test.go:
```diff
@@ -0,0 +1,96 @@
+		t.Skip("skipping test, zypper not found")
+	}
+
+	test.EnsurePrivilege(t)
```
`test.EnsurePrivilege(t)` should go on the first line of the func (niptick)

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper_test.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+func TestZypperPacker(t *testing.T) {
+
```
white line

ArangoGutierrez commented on src/pkg/build/sources/conveyorPacker_zypper_test.go:
```diff
@@ -0,0 +1,96 @@
+		t.Skip("skipping test, zypper not found")
+	}
+
+	test.EnsurePrivilege(t)
```
same as above

ikaneshiro committed:
removing white space and checking for priv at very start

ikaneshiro committed:
Merge branch 'master' of github.com:sylabs/singularity into zypper_dev_manual

ikaneshiro committed:
fixing comments

ikaneshiro committed:
Merge branch 'master' into zypper_dev_manual

ikaneshiro committed:
Merge branch 'master' into zypper_dev_manual

ikaneshiro committed:
Merge branch 'master' into zypper_dev_manual

GodloveD reviewed (dismissed):
I tried this on leap 42.3 and I got a panic.  Here's the stack trace.  Does this look like something I did or is there a bug?
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x9f6f99]

goroutine 1 [running]:
github.com/sylabs/singularity/src/pkg/build/sources.(*ZypperConveyorPacker).Get(0xc000482680, 0xc0000be640, 0x0, 0x0)
	/home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/sources/conveyorPacker_zypper.go:113 +0xbd9
github.com/sylabs/singularity/src/pkg/build.(*Build).Full(0xc0000be500, 0x1d, 0x7ffe614eee73)
	/home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/build.go:156 +0xda
github.com/sylabs/singularity/src/cmd/singularity/cli.glob..func4(0x143c800, 0xc0004764c0, 0x2, 0x2)
	/home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/build.go:106 +0x4d2
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).execute(0x143c800, 0xc0000300e0, 0x2, 0x2, 0x143c800, 0xc0000300e0)
	/home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x143f9e0, 0xc0004c5f88, 0x4078c0, 0xc00003a0b8)
	/home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).Execute(0x143f9e0, 0x0, 0x0)
	/home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:800 +0x2b
github.com/sylabs/singularity/src/cmd/singularity/cli.ExecuteSingularity()
	/home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/singularity.go:100 +0x2d
main.main()
	/home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli.go:16 +0x20
```

ikaneshiro committed:
fix error handling to not cause a panic

ikaneshiro committed:
remove unnecessary specific error message

ikaneshiro committed:
Merge remote-tracking branch 'upstream/master' into zypper_dev_manual

ikaneshiro committed:
use parser package to parse definition file

ikaneshiro committed:
change opensuse example to a newer version that will work with newer zypper versions

ikaneshiro committed:
Merge branch 'master' into zypper_dev_manual

jscook2345 commented on src/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,243 @@
+			"%s\n"+
+			"%s\n"+
+			"After creating the file, re-run the bootstrap.\n"+
+			"More info: https://github.com/singularityware/singularity/issues/241\n",
```
sylabs/singularity

ikaneshiro committed:
Merge remote-tracking branch 'upstream/master' into zypper_dev_manual

ikaneshiro committed:
Merge branch 'zypper_dev_manual' of github.com:ikaneshiro/singularity into zypper_dev_manual

ikaneshiro committed:
fix issue reference to use sylabs instead of singularityware

ikaneshiro committed:
bootstrap tweaks

bootstraps with --download-in-advance to allow for file conflict checks.
changes rpm db warning output to be cleaner.

ikaneshiro committed:
Merge remote-tracking branch 'upstream/master' into zypper_dev_manual

GodloveD committed:
Merge branch 'master' into zypper_dev_manual

ikaneshiro committed:
Merge remote-tracking branch 'upstream' into zypper_dev_manual

ikaneshiro committed:
fix newBundle() call

ikaneshiro committed:
use regex for mirrorurl parsing

ikaneshiro committed:
updating copyright year

ArangoGutierrez commented on internal/pkg/build/sources/conveyorPacker_zypper.go:
```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
2018-2019

ikaneshiro replied:
good catch

ArangoGutierrez commented on internal/pkg/build/sources/conveyorPacker_zypper_test.go:
```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
2018-2019

ikaneshiro committed:
better err handling

ikaneshiro committed:
pring string instead of bytes

ArangoGutierrez reviewed (approved):
LGTM

GodloveD committed:
Merge branch 'master' into zypper_dev_manual
